### ISSUE TITLE
Couple questions about training from scratch

### ISSUE THREAD

**Issue by rogercw** [CVLAB-Unibo/Real-time-self-adaptive-deep-stereo#8]
First, thanks for sharing the codes. I tried to train the network from the scratch, but I encountered some questions. Hope the authors could help me out.

1. Where to download the training dataset?
Based on "example_list.csv", I assume KITTI raw dataset was used to train the network. However, I did not find "proj_disp/groundtruth" folder in the data. Do you create them yourselves?

2. If I train with other dataset, do I need to manually divide the ground truth by 20?
In _https://github.com/CVLAB-Unibo/Real-time-self-adaptive-deep-stereo/issues/4#issuecomment-469774308_, it said the ground truth is scaled by -20. Does that mean I have to do the same thing for the other dataset manually as well?

3. I am confused by the the additional value scaling for the disparity. Would you mind explaining it a little bit more?
For example, why we need "/ scale[5]", instead of "x 2", in the end of https://github.com/CVLAB-Unibo/Real-time-self-adaptive-deep-stereo/blob/97ee32e0beff3141f279945186fc8de7f0a6c245/Nets/MadNet.py#L274? According to https://github.com/CVLAB-Unibo/Real-time-self-adaptive-deep-stereo/blob/97ee32e0beff3141f279945186fc8de7f0a6c245/Losses/loss_factory.py#L291, I assume the "real_disp_v6" is in the unit of the downscale resolution. Even though there is a scale argument for self._make_disp, that scale is not used. Therefore, I guess V6 and real_disp_v6 are in the same disparity unit. How to determine whether additional value scaling is needed or not when rescale to original resolution here? https://github.com/CVLAB-Unibo/Real-time-self-adaptive-deep-stereo/blob/97ee32e0beff3141f279945186fc8de7f0a6c245/Nets/MadNet.py#L362 May I know which part I might misunderstand?

4. What is the logic for the valid map in loss computation?
https://github.com/CVLAB-Unibo/Real-time-self-adaptive-deep-stereo/blob/97ee32e0beff3141f279945186fc8de7f0a6c245/Losses/loss_factory.py#L285 Is there any chance it could be `tf.logical_or(tf.equal(targets, 0), tf.greater(target, max_disp))` instead for the `tf.where` condition? Btw, when reading the disparity map, I think it might be needed to convert invalid value into 0, like INF in Middlebury dataset. Otherwise, the valid map mechanism might not work.

Please kindly let me know which question above are not clear enough. I will try to elaborate more. Thanks for your help.

**Comment by AlessioTonioni** [comment]
Hi!

1. Yes we create them converting the depth measurement to disparity knowing the camera parameters.

2. No The training code should already do it for you: the full resolution raw predictions of the network are multiplied by -20 before being compared to the groundtruth, therefore the network should learn to predict -true_dips/20.

3.  There might be a bug in the loss factory, let me investigate it a little bit more and then I'll update the answer.
 Anyway the idea is that at each scale we predict a scaled down version of the full res disparity, with disparity value equal to those that we would have at full res divided by 20. This trick helps balancing the loss values, but it should not be crucial.
 When going to the upper resolution we convert them to be meaningful disparity for that resolution, first multiplying them by 20 (i.e. going to the true disparity value) and then dividing them by the proper scale.

4. On many dataset we have groundtruth only for a subset of pixels, the valid map allows us to compute the loss only on those pixels ignoring all the others. 
Yes actually it could have been implemented also as a `tf.logical_or` followed by a cast to `tf.float32`, probably it is also more readable, I'll probably update the code to use boolean expression. 
You are right if your invalid pixels have disparity different from 0 you should add some lines of code to do the conversion.

**Comment by AlessioTonioni** [comment]
More on question number 3.

MADNet predicts full-scale disparities divided by 20 at each scale.  
The method `get_disparities()` returns a list of all the prediction already rescaled to full resolution. Therefore 
https://github.com/CVLAB-Unibo/Real-time-self-adaptive-deep-stereo/blob/97ee32e0beff3141f279945186fc8de7f0a6c245/Losses/loss_factory.py#L290
will always evaluate to 1 effectivelly training the network as we want.

**Comment by rogercw** [comment]
Hi @AlessioTonioni, thanks for the quick response! I think I have clearer picture now, and I will look into the codes again.  Regarding the training dataset, i.e. Q1 above, I have one followup question: Did you mean you just converting the depth measurement of KITTI dataset to disparity? If so, may I know how you got the depth measurement in the first place?

**Comment by AlessioTonioni** [comment]
> Did you mean you just converting the depth measurement of KITTI dataset to disparity? If so, may I know how you got the depth measurement in the first place?

Yes, we use the depth measurements available here: http://www.cvlibs.net/datasets/kitti/eval_depth.php?benchmark=depth_completion

> One additional question: Is there a particular reason why Scale 1 (downscaled by 2 in both axes) is skipped in MADNet? In other words, why the rescaled_prediction is directly scaled from Scale 2 (downscaled by 4 in both axes) result?

We do not extend the up-pyramid to half or full res mainly to keep a high framerate. On preliminary experiments with different variants of the architecture, we observed that the gain obtained by going up to Scale 2 was minimal compared to the loss in speed.

**Comment by rogercw** [comment]
Thanks for the information. It is helpful! If you already shared the disparity converting script somewhere, could you please point out where it is? If not, I will try to figure out how to do it myself. Thanks again for your time and help.

**Comment by meteorshowers** [comment]
@AlessioTonioni 
Great work!
could you share your the disparity converting script  on kitti？
Thanks a lot！
